### PR TITLE
Polis Integration

### PR DESCRIPTION
**Pull request by cronos-polis**
Experimental Polis Integration.
We launch btcpayserver only using Polis. Will setup for testing.

### CODE REVIEW

**Comment by NicolasDorier** [comment]
note that as DOGECOIN, merchant won't be able to use it because there is no source rate.

**Comment by cronos-polis** [comment]
What you mean with source rate?

**Comment by NicolasDorier** [comment]
If your merchant create an invoice of 10 USD, BTCpay has not way to convert in POLIS because  https://apiv2.bitcoinaverage.com/indices/global/ticker/short does not list the ticker.

**Comment by cronos-polis** [comment]
Ok. We can wait the merge then. Will try to add an API for coinmarketcap.

**Comment by lepipele** [comment]
@cronos-polis I've talked with @NicolasDorier about refactoring rates fetching logic. I'll hit you up on Slack with details... I do think it makes sense to offer Coinmaketcap API as alternative, but we want to do it in a way that makes it easy to plug in. Currently our merchants are used to Coinaverage which unfortunately doesn't have good niche support.

**Comment by NicolasDorier** [comment]
Actually I made https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Services/Rates/BTCPayRateProviderFactory.cs#L68

For being able later to use direct exchange API to access the rate instead of CoinAverage (or by using coinaverage as a fallback).

**Comment by lepipele** [comment]
Damn man, you already refactored it completely. I'll hit you up once I'm done with #122 wanna understand few design decisions on that.

**Comment by NicolasDorier** [comment]
Yes, it is still not perfect though, and also there is no way to select different exchange depending on the currency. Still work to do on this, but this is good first step.

**Comment by cronos-polis** [comment]
Wow, you guys are fast. @lepipele I'll contact you on slack.

**Comment by NicolasDorier** [comment]
Now the rate fetching is scriptable. You need to tell use though how rate should be calculated for POLIS in BTCPay out of the box.

**Comment by NicolasDorier** [comment]
See https://twitter.com/BtcpayServer/status/992032506944700418

**Comment by NicolasDorier** [comment]
mmh I need to refactor. Your exchange will be replaced when the list will be refreshed (every 5 min after boot)

**Comment by NicolasDorier** [comment]
Ok done, add you exchange like I did for CoinAverage

https://github.com/btcpayserver/btcpayserver/commit/26db946392927c3387cd9d263717f6fbf41bf054#diff-f0b4690b58d2394686d0d8ad65c8958bR98

and 

https://github.com/btcpayserver/btcpayserver/commit/26db946392927c3387cd9d263717f6fbf41bf054#diff-f0b4690b58d2394686d0d8ad65c8958bR78

**Comment by cronos-polis** [comment]
@NicolasDorier 
We added the exchange to BTCPayRateProviderFactory.cs
Seems that the cryptopia exchange is already supported to ExchangeSharp.
We added this line:
```
            DirectProviders.Add("cryptopia", new ExchangeSharpRateProvider("cryptopia", new ExchangeCryptopiaAPI(), false));
````

https://github.com/jjxtra/ExchangeSharp/blob/master/ExchangeSharp/API/Exchanges/ExchangeCryptopiaAPI.cs

File: BTCPayServer/Services/Rates/CoinAverageSettings.cs [hunk 1]
```diff
@@ -121,6 +121,7 @@ public CoinAverageSettings()
                 (DisplayName: "Quoine", Name: "quoine"),
                 (DisplayName: "BTC Markets", Name: "btcmarkets"),
                 (DisplayName: "Bitso", Name: "bitso"),
+                (DisplayName: "Cryptopia", Name: "cryptopia"),
```
Comment (NicolasDorier) [inline BTCPayServer/Services/Rates/CoinAverageSettings.cs]: remove this line

**Comment by NicolasDorier** [comment]
Remove the line I pointed out and then add it here https://github.com/btcpayserver/btcpayserver/commit/26db946392927c3387cd9d263717f6fbf41bf054#diff-f0b4690b58d2394686d0d8ad65c8958bR98

**Comment by cronos-polis** [comment]
Done

**Comment by NicolasDorier** [comment]
still need to add the line where I said.

File: BTCPayServer/BTCPayNetworkProvider.Polis.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+                NBXplorerNetwork = nbxplorerNetwork,
+                UriScheme = "polis",
+                DefaultRateRules = new[]
+                {
+                                "POLIS_X = POLIS_BTC * BTC_X",
+                                "POLIS_BTC = cryptopia(POLIS_BTC)"
+                },
+                CryptoImagePath = "imlegacy/polis.png",
+                DefaultSettings = BTCPayDefaultSettings.GetDefaultSettings(NetworkType),
+                CoinType = NetworkType == NetworkType.Mainnet ? new KeyPath("3'") : new KeyPath("1'")
```
Comment (NicolasDorier) [inline BTCPayServer/BTCPayNetworkProvider.Polis.cs]: incorrect KeyPath, ask to add yours on https://github.com/satoshilabs/slips/blob/master/slip-0044.md

**Comment by NicolasDorier** [comment]
I added cryptopia in my own commit on https://github.com/btcpayserver/btcpayserver/commit/449738414b1188691540c5fa0d7149dcac4cb1af, please rebase your changes

**Comment by cronos-polis** [comment]
@NicolasDorier Done we are waiting the PR from Keypath

**Comment by NicolasDorier** [comment]
very good !

**Comment by cronos-polis** [comment]
Hello @NicolasDorier Any news on this?
Is there anything we need to do?

**Comment by NicolasDorier** [comment]
sorry I forgot about it

**Comment by NicolasDorier** [comment]
POLIS support removed because of cryptopedia being hacked 884acdd and breaking btcpay tests

**Comment by eabz** [comment]
@NicolasDorier We will look towards the reintegration. 
Thanks.

**Comment by NicolasDorier** [comment]
@eabz I will not do it myself, so please make a PR to fix it.

**Pull request merged**